### PR TITLE
Add an endpoint to process ifname request label

### PR DESCRIPTION
**Pull request by nickolaev**

### CODE REVIEW

File: examples/icmp/k8s/simple-client.yaml [hunk 1]
```diff
@@ -19,4 +19,4 @@ metadata:
   name: "simple-client"
   namespace: default
   annotations:
-    ns.networkservicemesh.io: icmp-responder?app=simple-client
+    ns.networkservicemesh.io: icmp-responder?app=simple-client&ifname=responder0
```
Comment (edwarnicke) [inline examples/icmp/k8s/simple-client.yaml:22]: Could we use:

icmp-responder/responder0?app=simple-client

as speced out here:

https://github.com/networkservicemesh/networkservicemesh/issues/708

**Comment by edwarnicke** [comment]
It looks like all the pieces are in place such that

icmp-responder/responder0 

would work out of the box as intended.  SDK seems to respect the Intf when constructing its configuration.

**Comment by nickolaev** [comment]
@edwarnicke yes iet does, but that is for the interface name on the Client side, while here we are making the Client tell the Endpoont how to name the interface on the Endpoint side.

**Comment by edwarnicke** [comment]
How do you handle the collisions with more than one client asking for an iface name on the Endpoint side?

I'm also a little confused why the Client would care what the iface name is on the endpoint side... I suspect you can enlighten me with a use case :)

**Comment by nickolaev** [comment]
No protection, that is entirely Client's responsibility. Note that this is a completely custom Endpoint functionality specific for this use-case. It was requested by @networkop for implementing a set of Quagga routers in a mesh similar to what he did for his presentation:
https://static.sched.com/hosted_files/onseu19/76/ONS%20EU19%20K8s%20Netsim%20.pdf

and specifically this impressive demo:
https://github.com/networkop/k8s-topo#750-node-random-quagga-router-topology

**Comment by edwarnicke** [comment]
OK... but *why* does he need the *client* to dictate the NSEs choice of interface?

I can't quite garner that from the linked slides/demos.

**Comment by nickolaev** [comment]
Does it really matter *why*? I mean we can do this if we want to, therefore we show it.

**Comment by networkop** [comment]
Based on my understanding of NSM, the original idea is for NSE to deal with all the connection-level details and NSC is just a simple client. What I was trying to do is to use NSM for physical network simulations, where NSC and NSE simply represent two sides of the p2p link. The main goal is to connect pairs of interfaces with predictable naming. e.g. eth123 on NSC connects to eth456 on NSE. Since it's the NSC that initiates the connection request and it happens asynchronously (as pods get scheduled and spun up), the only way to guarantee that predictable naming is to pass it in the connection request.

**Comment by edwarnicke** [comment]
@networkop Welcome to the part!  So glad to have you involved :)

The thing I still don't follow is this why do the interface names have to algorithmically match in some way?

The NSC doesn't see the NSEs interface name, and the NSE doesn't see the NSC's interface name.
I'm confused about why they have to match :)

**Comment by networkop** [comment]
thank you @edwarnicke, I'm humbled to be here. 
One of the things that I do when simulating a physical network, is inject a startup network configuration. This network configuration can have references to interfaces (e.g. which interface to enable OSPF), so it's important for me to have predictable interface naming. Let's say i have the following topology:
IMG_20191017_150812
On R1, I have a configuration with static routes pointing to an interface (instead of a next-hop), in which case eth0 HAS to be connected to R2. And if we rely on NSE-side interface naming (e.g. R1's first interface is eth0, followed by eth1, eth2 an so on), this may not always be true since I can't guarantee the order in which R2 and R3 are going to come up. So the only way to guarantee that eth0 of R1 will always be connected to eth0 of R2 is for R2 (NSC) to pass this information to R1 (NSE) as a part of connection request.

**Comment by edwarnicke** [comment]
@networkop OK... starting to get clearer to me.  You are basically laying out a statically configured network topology... but with the dynamism of Pods running where ever they may run.  Is that right?  Sounds cool :)

Question... are you basically pushing a quagga config into each of your Router/Pods using ConfigMaps ?

**Comment by edwarnicke** [comment]
@networkop Oh... also... I totally get why you need the interfaces to be what you expect and connect to what you expect now... do you *need* the name eth0?  I ask... because most CNI plugins grab it, and so we *can't* grab it via NSM for a vWire (we don't interfere with CNI).

**Comment by networkop** [comment]
> @networkop OK... starting to get clearer to me. You are basically laying out a statically configured network topology... but with the dynamism of Pods running where ever they may run. Is that right? Sounds cool :)
> 
> Question... are you basically pushing a quagga config into each of your Router/Pods using ConfigMaps ?

yes to both questions. Quagga is just an example here, it can be any virtual appliance from any vendor. the idea is to be able to validate and test network changes before rolling them out in prod, so running configuration would come from real production devices with very minimal changes and get injected as a configmap.

And yeah, eth0 was just an example here. I understand that it will always be managed by the CNI plugin. It usually serves as an out-of-band management to remotely connect (ssh/telnet/netconf) into each individual device. So I'm only interested in connecting interfaces eth1+

**Comment by nickolaev** [comment]
Or even 'quagga0+' 😀

**Comment by edwarnicke** [comment]
OK... so you basically need NSM to give you a patch panel.

router-1 has ports [port1, port2, ... portN]
...
router-N has ports [port1, port2, ... portN]

and you want to be able to connect tuples
(router-n,port-m) <-> (router-q,port-r)

Is that about right?

**Comment by networkop** [comment]
that's exactly what I want, yes.

**Comment by edwarnicke** [comment]
OK.. perhaps something like:
```
apiVersion: networkservicemesh.io/v1alpha1
kind: NetworkService
metadata:
  name: netsim
spec:
  payload: IP
  matches:
    - match:
      sourceSelector:
        router: router1
        port: port1
      route:
        - destination:
          destinationSelector:
            router: route2
            port: port2
```

Which would connect a request for a vWire with labels:

{router=router1,port=port1} -> {router=router2,port=port2} ?

Router1 can be dumb as rocks, and just ask for 'netsim' with its labels as a Network Service... and if you decide you want to change to:

{router=router1,port=port1} -> {router=router3,port=port12} ?

you would just change the Network Service Definition... not the underlying router Pods.

Thoughts?

**Comment by networkop** [comment]
that sounds like a good idea, yes. Define both sides of the link as labels and apply them to the NSC pod.

**Comment by nickolaev** [comment]
Supressed by #72.

**Pull request closed without merging**